net/speedtest/speedtest_client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+)
+
+// StartClient dials the given address and starts a speedtest.
+// It returns any errors that come up in the tests.
+// It returns an error if the given test type isn't either download or upload.
+// If there are no errors in the test, it returns a slice of results.
+func StartClient(config TestConfig, host, port string) ([]Result, error) {
+	serverAddr, err := net.ResolveTCPAddr("tcp", host+":"+port)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := net.DialTCP("tcp", nil, serverAddr)
```

Comment:
You can use net.Dial instead, and then you don't have to manually ResolveTCPAddr first.

Also, use net.JoinHostPort rather than manual concatenation. That function will correctly handle IPv6 addrs.